File: problems/python/ParkingLotSystem/src/strategy/fee/flat_rate_fee_strategy.py
"""Flat rate fee strategy."""
from model.parking_ticket import ParkingTicket
from strategy.fee.fee_strategy import FeeStrategy


class FlatRateFeeStrategy(FeeStrategy):
    def __init__(self, rate_per_hour: float = 5.0):
        self._rate_per_hour = rate_per_hour

    def calculate_fee(self, ticket: ParkingTicket) -> float:
        hours = ticket.get_duration_in_hours()
        return max(0, hours * self._rate_per_hour)

File: problems/java/ParkingLotSystem/src/strategy/fee/FlatRateFeeStrategy.java
package strategy.fee;

import model.ParkingTicket;

public class FlatRateFeeStrategy implements FeeStrategy {

    private final double flatRatePerHour;

    public FlatRateFeeStrategy(double flatRatePerHour) {
        if (flatRatePerHour <= 0) {
            throw new IllegalArgumentException("flatRatePerHour must be positive");
        }
        this.flatRatePerHour = flatRatePerHour;
    }

    @Override
    public double calculateFee(ParkingTicket parkingTicket) {
        double totalHours = Math.ceil(parkingTicket.getDurationInHours());
        return totalHours * flatRatePerHour;
    }
}
